test/protocore/register_vote_internal.js

```diff
@@ -0,0 +1,189 @@
+      config.coconsensusAddress,
+      config.metachainId,
+      config.domainSeparator,
+      config.epochLength,
+      config.genesisKernelHeight,
+      config.genesisKernelHash,
+      config.genesisParentVoteMessageHash,
+      config.genesisSourceTransitionHash,
+      config.genesisSourceBlockHash,
+      config.genesisTargetBlockHash,
+      config.genesisSourceBlockNumber,
+      config.genesisTargetBlockNumber,
+      config.genesisProposedMetablockHeight,
+    );
+  });
+
+  contract('Negative Tests', async () => {
+    it('should fail if a finalisation status of a target checkpoint '
+      + 'is less than registered', async () => {
+      const v = await ProtocoreUtils.Validator.create();
+      const sig = await v.ecsign(Utils.getRandomHash());
+
+      await Utils.expectRevert(
+        config.protocore.registerVote(
+          Utils.getRandomHash(),
+          sig.r, sig.s, sig.v,
+        ),
+        'The given link status is at least reported.',
+      );
+    });
+    it('should fail if a link height inclusion principle is not kept', async () => {
+      const v = await ProtocoreUtils.Validator.create();
+
+      const {
+        voteMessageHash,
+      } = await ProtocoreUtils.proposeNonFinalisationLinkInternal(
+        config.protocore,
+        config.genesisVoteMessageHash,
+      );
+
+      await incrementMetablockHeight(config.protocore, config.coconsensusAddress);
+      await incrementMetablockHeight(config.protocore, config.coconsensusAddress);
+
+      const sig = await v.ecsign(voteMessageHash);
+
+      await Utils.expectRevert(
+        config.protocore.registerVote(
+          voteMessageHash,
+          sig.r, sig.s, sig.v,
+        ),
+        'Link height inclusion principle has surpassed.',
+      );
+    });
+    it('should fail if a validator has already voted for a height', async () => {
+      const openMetablockHeight = await config.protocore.openKernelHeight();
+      const nextMetablockHeight = openMetablockHeight.addn(1);
+
+      const v = await ProtocoreUtils.Validator.create();
+      await addToFVS(config.protocore, [v], nextMetablockHeight);
+
+      await incrementMetablockHeight(config.protocore, config.coconsensusAddress);
+
+      const {
+        voteMessageHash,
+      } = await ProtocoreUtils.proposeNonFinalisationLinkInternal(
+        config.protocore,
+        config.genesisVoteMessageHash,
+      );
+
+      const sig = await v.ecsign(voteMessageHash);
+
+      await config.protocore.registerVote(
+        voteMessageHash,
+        sig.r, sig.s, sig.v,
+      );
+
+      await Utils.expectRevert(
+        config.protocore.registerVote(
+          voteMessageHash,
+          sig.r, sig.s, sig.v,
+        ),
+        'Validator vote cannot be registered twice in FVS votes at the same height.',
+      );
+    });
+  });
+
+  contract('Positive Tests', async () => {
+    it('checks that a vote of a validator is not double-counting', async () => {
```

Comment:
OK to leave empty, but write `it.skip(' .. ')`

Reply:
added